### PR TITLE
[feat][broker][PIP-195] Support internal cursor properties - part4

### PR DESCRIPTION
**Pull request by coderzc**
Master Issue: #16763 

### Motivation

Now, the cursor properties support overall update and modifying a single value concurrently by #17164, but `setCursorProperties` will all key-value is removed before update properties, this will cause internal property loss, so use a special prefix to prevent internal properties.

More context see: https://github.com/apache/pulsar/pull/17164#issuecomment-1236872277, https://github.com/apache/pulsar/pull/17164#issuecomment-1241523618, https://github.com/apache/pulsar/pull/17164#issuecomment-1241536546

### Modifications

Use a special prefix to prevent internal properties.

### Verifying this change

- [x] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Documentation

- [ ] `doc-required` 
(Your PR needs to update docs and you will update later)

- [x] `doc-not-needed` 
(Please explain why)

- [ ] `doc` 
(Your PR contains doc changes)

- [ ] `doc-complete`
(Docs have been already added)

Matching PR in forked repository
PR in forked repository: https://github.com/coderzc/pulsar/pull/4

### CODE REVIEW

**Comment by coderzc** [comment]
@lhotari @eolivelli @Jason918 @codelipenghui @mattisonchao PTAL.

**Comment by lhotari** [comment]
the tests failed in Broker Group 3 because of OOM, I filed #17714 . It's most likely not caused by the changes in this PR.

**Comment by coderzc** [comment]
@lhotari @eolivelli @Jason918 @codelipenghui @mattisonchao PTAL

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java [hunk 1]
```diff
@@ -371,7 +373,28 @@ public void operationFailed(MetaStoreException e) {
...
     public CompletableFuture<Void> setCursorProperties(Map<String, String> cursorProperties) {
-        return computeCursorProperties(lastRead -> cursorProperties);
+        Map<String, String> newProperties =
+                cursorProperties == null ? new HashMap<>() : new HashMap<>(cursorProperties);
+
+        // Prohibit setting of internal properties
+        Set<String> keys = newProperties.keySet();
+        for (String key : keys) {
+            if (key.startsWith(CURSOR_INTERNAL_PROPERTY_PREFIX)) {
+                throw new IllegalArgumentException(
```
Comment (mattisonchao) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java:383]: I think returning failed `CompletableFuture` will be more make sense for the caller.

File: managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/ManagedCursorPropertiesTest.java [hunk 2]
```diff
@@ -232,6 +235,31 @@ void testUpdateCursorProperties() throws Exception {
...
         factory2.shutdown();
+
+        // Create a new factory to force a managed ledger close and recovery
+        ManagedLedgerFactory factory3 = new ManagedLedgerFactoryImpl(metadataStore, bkc);
+        // Reopen the managed ledger
+        ledger = factory3.open("testUpdateCursorProperties", new ManagedLedgerConfig());
+        c1 = ledger.openCursor("c1");
+
+        c1.putCursorProperty(CURSOR_INTERNAL_PROPERTY_PREFIX + "test", "test").get();
+        c1.putCursorProperty("custom4", "custom4").get();
```
Comment (mattisonchao) [inline managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/ManagedCursorPropertiesTest.java:246]: If we call the `CompletableFuture#get` method, it's better to give the test a timeout.

**Review by mattisonchao — APPROVED** [review]
It appears to have a minor compatibility issue - if the user has previously used a format like `CURSOR_INTERNAL_PROPERTY_PREFIX` in cursor properties. I think it's a small probability event though.

**Comment by coderzc** [comment]
@lhotari @eolivelli @Jason918 @codelipenghui PTAL

**Review by Jason918 — APPROVED** [review]
LGTM

**Review by codelipenghui — APPROVED** [review]

**Pull request merged**